src/Native/LDANative/model_block.cpp

```diff
@@ -0,0 +1,467 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+#include <fstream>
+#include <iostream>
+#include <vector>
+#include "utils.hpp"
+#include <stdlib.h>
+#include "model_block.h"
+#include "lda_document.h"
+
+namespace lda
+{
+    int64_t upper_bound(int64_t x)
+    {
+        if (x == 0)
+        {
+            return 0;
+        }
+        int64_t shift = 0;
+        int64_t y = 1;
+        x--;
+        while (x)
+        {
+            x = x >> 1;
+            y = y << 1;
+            ++shift;
+        }
+        return y;
+    }
+
+    int32_t align64(int32_t size)
+    {
+        if (size % 64 == 0)
+        {
+            return size;
+        }
+        else
+        {
+            size = 64 * (size / 64) + 64;
+            return size;
+        }
+    }
+
+
+    LDAModelBlock::LDAModelBlock()
+        : dict_(nullptr),
+        num_vocabs_(0),
+        mem_block_size_(0),
+        mem_block_(nullptr),
+        alias_mem_block_size_(0),
+        alias_mem_block_(nullptr)
+    {
+    }
+    LDAModelBlock::~LDAModelBlock()
+    {
+        Clear();
+    }
+
+    void LDAModelBlock::Clear()
+    {
+        if (dict_)
+        {
+            delete[]dict_;
+            dict_ = nullptr;
+        }
+        if (mem_block_)
+        {
+            delete[]mem_block_;
+            mem_block_ = nullptr;
+        }
+        if (alias_mem_block_)
+        {
+            delete[]alias_mem_block_;
+            alias_mem_block_ = nullptr;
+        }
+
+        num_vocabs_ = -1;
+        num_topics_ = -1;
+
+        mem_block_size_ = 0;
+        alias_mem_block_size_ = 0;
+    }
+
+    void LDAModelBlock::Init(int32_t num_vocabs, int32_t num_topics, int64_t nonzero_num)
+    {
+        num_vocabs_ = num_vocabs;
+        num_topics_ = num_topics;
+
+        dict_ = new WordEntry[num_vocabs_];
+        for (int i = 0; i < num_vocabs_; ++i)
+        {
+            // This warning is a false positive. Supressing it similar to the existing one on Line 140 below.
+#pragma warning(suppress: 6386)
+            dict_[i].is_dense_ = 0;
+            dict_[i].is_alias_dense_ = 0;
+        }
+
+        // mem_block_size_ = nonzero_num * 2;        // TODO: power of 2, load factor of 2, key-value of 2
+        mem_block_size_ = 2 * upper_bound(load_factor_ * nonzero_num);
+        alias_mem_block_size_ = nonzero_num * 3; 
+
+        mem_block_ = new int32_t[mem_block_size_]();                // NOTE(jiyuan) : force to initialize the values to be zero
```

Comment:
>jiyuan [](start = 76, length = 6)

all this aliases

Reply:
#Resolved

Reply:
#Resolved

---
In reply to: [197899749](https://github.com/dotnet/machinelearning/pull/377#discussion_r197899749) [](ancestors = 197899749)